modules/core/src/quaternion.cpp

```diff
@@ -0,0 +1,742 @@
+    return c2.log();
+    //return cv::log(*this + cv::power(*this * *this - Quat<T>(1,0,0,0), 0.5));
+}
+/*
+
+template <typename T>
+inline Quat<T> cv::atanh(const Quat<T> &q)
+{
+    return q.atanh();
+}
+*/
+
+template <typename T>
+inline Quat<T> Quat<T>::atanh() const
+{
+    Quat<T> ident(1, 0, 0, 0);
+    Quat<T> c1 = (ident + *this).log();
+    Quat<T> c2 = (ident - *this).log();
+    return 1 / 2 * (c1 - c2);
+    //return 1/2 * (cv::log(ident + *this) - cv::log(ident - *this));
+}
+/*
+
+template <typename T>
+inline Quat<T> cv::asin(const Quat<T> &q)
+{
+    return q.asin();
+}
+*/
+
+template <typename T>
+inline Quat<T> Quat<T>::asin() const
+{
+    Quat<T> v(0, x, y, z);
+    T vNorm = v.norm();
+    return -v / vNorm * (*this * v / vNorm).asinh();
+}
+/*
+
+template <typename T>
+inline Quat<T> cv::acos(const Quat<T> &q)
+{
+    return q.acos();
+}
+*/
+
+template <typename T>
+inline Quat<T> Quat<T>::acos() const
+{
+    Quat<T> v(0, x, y, z);
+    T vNorm = v.norm();
+    return -v / vNorm * acosh();
+}
+/*
+
+template <typename T>
+inline Quat<T> cv::atan(const Quat<T> &q)
+{
+    return q.atan();
+}
+*/
+template <typename T>
+inline Quat<T> Quat<T>::atan() const
+{
+    Quat<T> v(0, x, y, z);
+    T vNorm = v.norm();
+    return -v / vNorm * (*this * v / vNorm).atanh();
+}
+
+template <typename T>
+inline T Quat<T>::getAngle() const
+{
+    return 2 * std::acos(w / norm());
+}
+
+template <typename T>
+inline cv::Vec<T, 3> Quat<T>::getAxis() const
+{
+    T angle = getAngle();
+    if (abs(std::sin(angle / 2)) < EPS)
+        return cv::Vec<T, 3> {x, y, z}; // TBD
+    return cv::Vec<T, 3> {x, y, z} / (norm() * std::sin(angle / 2));
+}
+
+template <typename T>
+cv::Mat Quat<T>::toRotMat4x4() const
+{
+    T dotVal = dot(*this);
+    cv::Matx<T, 4, 4> R{
+         dotVal, 0                           , 0                           , 0,
+         0     , dotVal - 2 * (y * y + z * z),  2 * (x * y + w * z)        , 2 * (x * z - w * y),
+         0     , 2 * (x * y - w * z)         , dotVal - 2 * (x * x + z * z), 2 * (y * z + w * x),
+         0     , 2 * (x * z + w * y)         , 2 * (y * z - w * x)         , dotVal - 2 * (x * x + y * y)};
+    return cv::Mat(R).t();
+}
+
+template <typename T>
+cv::Mat Quat<T>::toRotMat3x3() const
+{
+    assertNormal();
+    cv::Matx<T, 3, 3> R{
+          1 - 2 * (y * y + z * z), 2 * (x * y + w * z)    , 2 * (x * z - w * y),
+          2 * (x * y - w * z)    , 1 - 2 * (x * x + z * z), 2 * (y * z + w * x),
+          2 * (x * z + w * y)    , 2 * (y * z - w * x)    , 1 - 2 * (x * x + y * y)};
+    return cv::Mat(R).t();
+}
+
+template <typename T>
+cv::Vec<T, 4> Quat<T>::toVec() const
+{
+    return cv::Vec<T, 4>{w, x, y, z};
+}
+
+template <typename T>
+Quat<T> Quat<T>::lerp(const Quat<T> &q0, const Quat<T> &q1, const T t)
+{
+    return (1 - t) * q0 + t * q1;
+}
+
+template <typename T>
+Quat<T> Quat<T>::slerp(Quat<T> &q0, Quat<T> &q1, const T t, bool assumeUnit)
```

Comment:
Input arguments should be const references

Reply:
At first I considered that q0 and q1 might be changed because of the choice of direction interpolation. but it will also change the user's variable.

In my latest code, I add the const and create a new quaternion in function like this
```
template <typename T>
Quat<T> Quat<T>::slerp(const Quat<T> &q0, const Quat<T> &q1, const T t, bool assumeUnit)
{
    Quat<T> v0(q0);
    Quat<T> v1(q1);
    ....
}
```
and then to return the calculation of v0 and v1.